File: test/BridgeNetTests/TypeScript/App1/4_Functions.cs
namespace Functions
{
    public delegate j j(MiddleBit x);

    public class MiddleBit
    {
        public j fn;
    }

    public class Parameters
    {
        // TODO #292
        public int GetSomething(int i = 5)
        {
            return i;
        }

        public string Join(params int[] numbers)
        {
            string s = string.Empty;
            for (int i = 0; i < numbers.Length; i++)
            {
                s = s + numbers[i];
            }

            return s;
        }
    }

    public class Delegates
    {
        public delegate void VoidDelegate();

        public delegate void StringDelegate(string s);

        public delegate int StringDelegateIntResult(string s);
    }

    // TODO
    public interface DelegateInterface
    {
        event Delegates.VoidDelegate MethodVoidDelegate;

        event Delegates.StringDelegate MethodStringDelegate;

        event Delegates.StringDelegateIntResult MethodStringDelegateIntResult;
    }

    public class DelegateClass
    {
        public Delegates.VoidDelegate MethodVoidDelegate;
        public Delegates.StringDelegate MethodStringDelegate;
        public Delegates.StringDelegateIntResult MethodStringDelegateIntResult;
    }
}

File: test/BridgeNetTests/Runner/TypeScript/Batch1/4_Functions.js
/// <reference path="..\..\Runner\resources\qunit\qunit.d.ts" />
/// <reference path="..\..\Runner\TypeScript\App1\bridge.d.ts" />
/// <reference path="..\..\Runner\TypeScript\App1\Functions.d.ts" />
"use strict";
QUnit.module("TypeScript - Functions");
QUnit.test("Parameters", function (assert) {
    var func = new Functions.Parameters();
    // TODO Bridge/#292
    // assert.deepEqual(func.getSomething(), 5, "Default parameter #292");
    //function buildName(firstName: string, lastName = "Smith") {
    //    // JavaScript added for the default parameter
    //    // if (typeof lastName === "undefined") { lastName = "Smith"; }
    //    return firstName + " " + lastName;
    //}
    //var result1 = buildName("Bob");
    // #293
    assert.deepEqual(func.Join([1, 2, 3]), "123", "params argument becomes Array #293");
});
QUnit.test("Function types", function (assert) {
    var d = new Functions.DelegateClass();
    var ds;
    var di;
    d.MethodVoidDelegate = function () { return di = 7; };
    d.MethodVoidDelegate();
    assert.deepEqual(di, 7, "methodVoidDelegate");
    d.MethodStringDelegate = function (s) { return ds = s; };
    d.MethodStringDelegate("Privet");
    assert.deepEqual(ds, "Privet", "methodStringDelegate");
    d.MethodStringDelegateIntResult = function (s) { return s.length; };
    assert.deepEqual(d.MethodStringDelegateIntResult("Privet"), 6, "methodStringDelegateIntResult");
});
